### PR TITLE
Big Data Solution Nearsoft Academy

### PR DESCRIPTION
**Pull request by pmduaree**
Here is the solution of the big data problem made by the intern's team.
The program takes 9 minutes cleaning the file and the recommendations.

-Interns

### CODE REVIEW

File: src/test/java/nearsoft/academy/bigdata/recommendation/TextReducer.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.zip.GZIPInputStream;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import org.apache.mahout.cf.taste.common.TasteException;
+import org.apache.mahout.cf.taste.impl.model.MemoryIDMigrator;
+ 
+public class TextReducer {
```
Comment (rilopez) [inline src/test/java/nearsoft/academy/bigdata/recommendation/TextReducer.java:14]: Wowww, this is really awesome, thanks for going the extra mile

File: src/test/java/nearsoft/academy/bigdata/recommendation/TextReducer.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        write("\n");
+        numberOfLines++;
+    } 
+    //number of lines = number of reviews
+    public int getNumberOfLines()
+    {
+        return numberOfLines;
+    }
+    //creates a substring using only indexOf and substring
+    private String substr(String line, char limit)
```
Comment (rilopez) [inline src/test/java/nearsoft/academy/bigdata/recommendation/TextReducer.java:106]: this method should have a better name  and it should say clearly what it does. so far it seem that this method gets the value part of  a colon separated key/value  pair.

File: src/test/java/nearsoft/academy/bigdata/recommendation/TextReducer.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        return line.substring(line.indexOf(limit)+1, line.length());
+    } 
+    //writes to the file
+    private void write(String line) throws IOException
+    {
+        writer.write(line);
```
Comment (rilopez) [inline src/test/java/nearsoft/academy/bigdata/recommendation/TextReducer.java:113]: This method doesn't make much sense right now. in some cases it's good to have wrapping methods like this for easy testing or mocking but here you are doing neither

File: src/test/java/nearsoft/academy/bigdata/recommendation/TextReducer.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            //read the next line (should be the product id or null)
+            line = is.readLine();
+        }
+        is.close();
+        writer.close();
+        
+    }
+
+    // clean the file
+    private void cleanBlock(String userId, String productId, String score) throws IOException
```
Comment (rilopez) [inline src/test/java/nearsoft/academy/bigdata/recommendation/TextReducer.java:78]: 1 ) this should have a better name , cleanBlock is not descritive 
2) it seems that its doing too much (writing a comma separated line, mapping the ids, incrementing the counters)

File: pom.xml [hunk 5]
```diff
@@ -26,5 +45,13 @@
           <version>4.7</version>
           <scope>test</scope>
       </dependency>
+      <dependency>
```
Comment (rilopez) [inline pom.xml:48]: Why are you duplicating the mahout dependency?

File: src/test/java/nearsoft/academy/bigdata/recommendation/MovieRecommender.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.mahout.cf.taste.model.DataModel;
+import org.apache.mahout.cf.taste.impl.recommender.GenericUserBasedRecommender;
+import org.apache.mahout.cf.taste.common.TasteException;
+import org.apache.mahout.cf.taste.recommender.UserBasedRecommender;
+import org.apache.mahout.cf.taste.recommender.RecommendedItem;
+import org.apache.mahout.cf.taste.recommender.Recommender;
+import org.apache.mahout.cf.taste.impl.model.MemoryIDMigrator;
+import org.apache.mahout.cf.taste.impl.recommender.CachingRecommender;
+
+public  class MovieRecommender {
```
Comment (rilopez) [inline src/test/java/nearsoft/academy/bigdata/recommendation/MovieRecommender.java:21]: This file should be on the main source tree not into the test sources tree

File: pom.xml (hunk 5 above)
Comment (charlierodarte) [inline pom.xml]: It seems that for unit testing they're using a different version. Will have to look further into it, though.

**Comment by charlierodarte** [comment]
I've reviewed this code as if it were SWF code. I see 1 big flaw. I'll explain:

The problem statement can be summarised in feeding the data to the PearsonCorrelationSimilarity object. It takes a 'model' (a DataModel, that is) as its argument. 

Your solution takes the file, processes it, dumps it into a different file and then lets Mahout read the 2nd file. See the problem? The file is read & written one too many times. No file writing should occur & only one file should be read, only once, not twice. 

Quickly skimming through Mahout's documentation it seems that a better solution would have been subclassing the FileDataModel class & override either the processLine or processLineWithoutId methods. Since Mahout is open source, you could have even taken a peek at the current implementations as to provide a nicer approach that does the mapping that you require, on-the-fly. 

That issue aside, the "temp.csv" file is nefarious. Java provides a mechanism through the File class to create temporary files. The MovieRecommender class should NOT expect a "temp.csv" file to have been created by the FileReducer class. The FileReducer "clean" method should return a File object, that is, a reference to the file that was just created. The method should be renamed, too, "clean" is not clear at all. Furthermore, the "temp.csv" file is never destroyed! That's also an error. Destruction of the file should have taken place in an "@After" annotated method.

File: src/test/java/nearsoft/academy/bigdata/recommendation/MovieRecommender.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    //global variables
+    public DataModel model ;
+    public UserBasedRecommender recommender;
+    public UserSimilarity similarity;
+    public UserNeighborhood neighborhood;
+    public Recommender cachingRecommender;
+    public TextReducer textReducer;
+    private MemoryIDMigrator memoryIDMigrator;
+    //constructor
```
Comment (charlierodarte) [inline src/test/java/nearsoft/academy/bigdata/recommendation/MovieRecommender.java:30]: Remove this comment. It doesn't add any value to the code.

File: src/test/java/nearsoft/academy/bigdata/recommendation/MovieRecommender.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.mahout.cf.taste.impl.recommender.GenericUserBasedRecommender;
+import org.apache.mahout.cf.taste.common.TasteException;
+import org.apache.mahout.cf.taste.recommender.UserBasedRecommender;
+import org.apache.mahout.cf.taste.recommender.RecommendedItem;
+import org.apache.mahout.cf.taste.recommender.Recommender;
+import org.apache.mahout.cf.taste.impl.model.MemoryIDMigrator;
+import org.apache.mahout.cf.taste.impl.recommender.CachingRecommender;
+
+public  class MovieRecommender {
+    //global variables
```
Comment (charlierodarte) [inline src/test/java/nearsoft/academy/bigdata/recommendation/MovieRecommender.java:22]: Remove this comment. It doesn't add any value to the code.

**Pull request closed without merging**